File: core/src/main/java/com/vzome/core/viewing/Lights.java
/*
 * Created on Jul 22, 2004
 */
package com.vzome.core.viewing;

import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import javax.vecmath.Vector3f;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.vzome.core.model.Color;
import com.vzome.xml.DomUtils;

/**
 * This is really a SceneModel
 * @author Scott Vorthmann
 *
 */
public class Lights implements Iterable<Lights.DirectionalLight>
{
    public static class DirectionalLight
    {
        public DirectionalLight( Vector3f direction, Color color )
        {
            super();
            this.direction = direction;
            this.color = color;
        }

        public Vector3f direction;
        public Color color;
    }

    private final PropertyChangeSupport pcs = new PropertyChangeSupport( this );

    public void addPropertyListener( PropertyChangeListener listener )
    {
        pcs .addPropertyChangeListener( listener );
    }

    public void removePropertyListener( PropertyChangeListener listener )
    {
        pcs .removePropertyChangeListener( listener );
    }

    public void setProperty( String cmd, Object value )
    {
        if ( "backgroundColor".equals( cmd ) ) {
            this .backgroundColor = new Color( Integer .parseInt( (String) value, 16 ) );
            pcs .firePropertyChange( cmd, null, value );
        }
    }

    protected final List<DirectionalLight> directionalLights = new ArrayList<>(3);

    protected Color mAmbientLightColor;

    private Color backgroundColor;


    public Lights()
    {
        super();
    }


    public Lights( Lights prototype )
    {
        this();

        this .backgroundColor = prototype .backgroundColor;
        this .mAmbientLightColor = prototype .mAmbientLightColor;
        for ( int i = 0; i < prototype .directionalLights .size(); i++ ) {
            addDirectionLight( prototype .directionalLights .get(i) );
        }
    }


    public Lights( Element element )
    {
        this();
        String str = element .getAttribute( "background" );
        this .backgroundColor = Color .parseColor( str );
        str = element .getAttribute( "ambientLight" );
        this .mAmbientLightColor = Color .parseColor( str );
        NodeList nodes = element .getChildNodes();
        for ( int i = 0; i < nodes .getLength(); i++ ) {
            Node node = nodes .item( i );
            if ( node instanceof Element ) {
                Element viewElem = (Element) node;
                str = viewElem .getAttribute( "color" );
                Color color = Color .parseColor( str );
                Vector3f pos = new Vector3f( 
                        Float .parseFloat( viewElem .getAttribute( "x" ) ),  
                        Float .parseFloat( viewElem .getAttribute( "y" ) ),  
                        Float .parseFloat( viewElem .getAttribute( "z" ) )
                        );
                addDirectionLight( new DirectionalLight( pos, color ) );
            }
        }
    }

    public int size()
    {
        return this .directionalLights .size();
    }

    public final void addDirectionLight( DirectionalLight light )
    {
        directionalLights .add( light );
    }

    public void setAmbientColor( Color color )
    {
        mAmbientLightColor = color;
    }

    public Color getAmbientColor()
    {
        return mAmbientLightColor;
    }
    
    public void getDirectionalLights()
    {
        
    }

    public Color getDirectionalLight( int i, Vector3f direction )
    {
        DirectionalLight light = this .directionalLights .get( i );
        direction .set( light .direction );
        return light .color;
    }

    public Color getBackgroundColor()
    {
        return this .backgroundColor;
    }

    public void setBackgroundColor( Color color )
    {
        this .backgroundColor = color;
    }

    public Element getXml( Document doc )
    {
        Element result = doc .createElement( "sceneModel" );
        DomUtils .addAttribute( result, "ambientLight", mAmbientLightColor .toString() );
        DomUtils .addAttribute( result, "background", backgroundColor .toString() );
        for ( int i = 0; i < this .directionalLights .size(); i++ ) {
            DirectionalLight light = this .directionalLights .get( i );
            Element child = doc .createElement( "directionalLight" );
            DomUtils .addAttribute( child, "x", Float .toString( light .direction .x ) );
            DomUtils .addAttribute( child, "y", Float .toString( light .direction .y ) );
            DomUtils .addAttribute( child, "z", Float .toString( light .direction .z ) );
            DomUtils .addAttribute( child, "color", light .color .toString() );
            result .appendChild( child );
        }
        return result;
    }

    @Override
    @JsonGetter( "directionalLights" )
    public Iterator<DirectionalLight> iterator()
    {
        return this .directionalLights .iterator();
    }

    public void addDirectionLight( Color color, Vector3f dir )
    {
        this .addDirectionLight( new DirectionalLight( dir, color ) );
    }
}

File: online/src/worker/legacy/from-java/com/vzome/core/viewing/Lights.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { Color } from "../construction/Color.js";
import { RealVector } from "../math/RealVector.js";
import { DomUtils } from "../../xml/DomUtils.js";
import { PropertyChangeListener } from "../../../../java/beans/PropertyChangeListener.js";
import { PropertyChangeSupport } from "../../../../java/beans/PropertyChangeSupport.js";
import { Document } from "../../../../org/w3c/dom/Document.js";
import { Element } from "../../../../org/w3c/dom/Element.js";
import { Node } from "../../../../org/w3c/dom/Node.js";
import { NodeList } from "../../../../org/w3c/dom/NodeList.js";

/**
 * This is really a SceneModel
 * @author Scott Vorthmann
 * @param {Lights} prototype
 * @class
 */
export class Lights implements java.lang.Iterable<Lights.DirectionalLight> {
    /*private*/ pcs: PropertyChangeSupport;

    public addPropertyListener(listener: PropertyChangeListener) {
        this.pcs.addPropertyChangeListener$java_beans_PropertyChangeListener(listener);
    }

    public removePropertyListener(listener: PropertyChangeListener) {
        this.pcs.removePropertyChangeListener$java_beans_PropertyChangeListener(listener);
    }

    public setProperty(cmd: string, value: any) {
        if ("backgroundColor" === cmd){
            this.backgroundColor = new Color(javaemul.internal.IntegerHelper.parseInt(<string>value, 16));
            this.pcs.firePropertyChange$java_lang_String$java_lang_Object$java_lang_Object(cmd, null, value);
        }
    }

    directionalLights: java.util.List<Lights.DirectionalLight>;

    mAmbientLightColor: Color;

    /*private*/ backgroundColor: Color;

    public constructor(prototype?: any) {
        if (((prototype != null && prototype instanceof <any>Lights) || prototype === null)) {
            let __args = arguments;
            {
                let __args = arguments;
                if (this.mAmbientLightColor === undefined) { this.mAmbientLightColor = null; } 
                if (this.backgroundColor === undefined) { this.backgroundColor = null; } 
                this.pcs = new PropertyChangeSupport(this);
                this.directionalLights = <any>(new java.util.ArrayList<any>(3));
            }
            (() => {
                this.backgroundColor = prototype.backgroundColor;
                this.mAmbientLightColor = prototype.mAmbientLightColor;
                for(let i: number = 0; i < prototype.directionalLights.size(); i++) {{
                    this.addDirectionLight$com_vzome_core_viewing_Lights_DirectionalLight(prototype.directionalLights.get(i));
                };}
            })();
        } else if (((prototype != null && (prototype.constructor != null && prototype.constructor["__interfaces"] != null && prototype.constructor["__interfaces"].indexOf("org.w3c.dom.Element") >= 0)) || prototype === null)) {
            let __args = arguments;
            let element: any = __args[0];
            {
                let __args = arguments;
                if (this.mAmbientLightColor === undefined) { this.mAmbientLightColor = null; } 
                if (this.backgroundColor === undefined) { this.backgroundColor = null; } 
                this.pcs = new PropertyChangeSupport(this);
                this.directionalLights = <any>(new java.util.ArrayList<any>(3));
            }
            (() => {
                let str: string = element.getAttribute("background");
                this.backgroundColor = Color.parseColor(str);
                str = element.getAttribute("ambientLight");
                this.mAmbientLightColor = Color.parseColor(str);
                const nodes: NodeList = element.getChildNodes();
                for(let i: number = 0; i < nodes.getLength(); i++) {{
                    const node: Node = nodes.item(i);
                    if (node != null && (node.constructor != null && node.constructor["__interfaces"] != null && node.constructor["__interfaces"].indexOf("org.w3c.dom.Element") >= 0)){
                        const viewElem: Element = <Element><any>node;
                        str = viewElem.getAttribute("color");
                        const color: Color = Color.parseColor(str);
                        const pos: RealVector = new RealVector(javaemul.internal.FloatHelper.parseFloat(viewElem.getAttribute("x")), javaemul.internal.FloatHelper.parseFloat(viewElem.getAttribute("y")), javaemul.internal.FloatHelper.parseFloat(viewElem.getAttribute("z")));
                        this.addDirectionLight$com_vzome_core_viewing_Lights_DirectionalLight(new Lights.DirectionalLight(pos, color));
                    }
                };}
            })();
        } else if (prototype === undefined) {
            let __args = arguments;
            if (this.mAmbientLightColor === undefined) { this.mAmbientLightColor = null; } 
            if (this.backgroundColor === undefined) { this.backgroundColor = null; } 
            this.pcs = new PropertyChangeSupport(this);
            this.directionalLights = <any>(new java.util.ArrayList<any>(3));
        } else throw new Error('invalid overload');
    }

    public size(): number {
        return this.directionalLights.size();
    }

    public addDirectionLight$com_vzome_core_viewing_Lights_DirectionalLight(light: Lights.DirectionalLight) {
        this.directionalLights.add(light);
    }

    public setAmbientColor(color: Color) {
        this.mAmbientLightColor = color;
    }

    public getAmbientColor(): Color {
        return this.mAmbientLightColor;
    }

    public getAmbientColorWeb(): string {
        return this.mAmbientLightColor.toWebString();
    }

    public getDirectionalLights() {
    }

    public getBackgroundColor(): Color {
        return this.backgroundColor;
    }

    public getBackgroundColorWeb(): string {
        return this.backgroundColor.toWebString();
    }

    public setBackgroundColor(color: Color) {
        this.backgroundColor = color;
    }

    public getXml(doc: Document): Element {
        const result: Element = doc.createElement("sceneModel");
        DomUtils.addAttribute(result, "ambientLight", this.mAmbientLightColor.toString());
        DomUtils.addAttribute(result, "background", this.backgroundColor.toString());
        for(let i: number = 0; i < this.directionalLights.size(); i++) {{
            const light: Lights.DirectionalLight = this.directionalLights.get(i);
            const child: Element = doc.createElement("directionalLight");
            DomUtils.addAttribute(child, "x", /* toString */(''+(light.direction.x)));
            DomUtils.addAttribute(child, "y", /* toString */(''+(light.direction.y)));
            DomUtils.addAttribute(child, "z", /* toString */(''+(light.direction.z)));
            DomUtils.addAttribute(child, "color", light.color.toString());
            result.appendChild(child);
        };}
        return result;
    }

    /**
     * 
     * @return {*}
     */
    public iterator(): java.util.Iterator<Lights.DirectionalLight> {
        return this.directionalLights.iterator();
    }

    public addDirectionLight$com_vzome_core_construction_Color$com_vzome_core_math_RealVector(color: Color, dir: RealVector) {
        this.addDirectionLight$com_vzome_core_viewing_Lights_DirectionalLight(new Lights.DirectionalLight(dir, color));
    }

    public addDirectionLight(color?: any, dir?: any) {
        if (((color != null && color instanceof <any>Color) || color === null) && ((dir != null && dir instanceof <any>RealVector) || dir === null)) {
            return <any>this.addDirectionLight$com_vzome_core_construction_Color$com_vzome_core_math_RealVector(color, dir);
        } else if (((color != null && color instanceof <any>Lights.DirectionalLight) || color === null) && dir === undefined) {
            return <any>this.addDirectionLight$com_vzome_core_viewing_Lights_DirectionalLight(color);
        } else throw new Error('invalid overload');
    }

    public getDirectionalLightVector(i: number): RealVector {
        const light: Lights.DirectionalLight = this.directionalLights.get(i);
        return new RealVector(light.direction.x, light.direction.y, light.direction.z);
    }

    public getDirectionalLightColor(i: number): Color {
        const light: Lights.DirectionalLight = this.directionalLights.get(i);
        return light.color;
    }
}
Lights["__class"] = "com.vzome.core.viewing.Lights";
Lights["__interfaces"] = ["java.lang.Iterable"];



export namespace Lights {

    export class DirectionalLight {
        public constructor(direction: RealVector, color: Color) {
            if (this.direction === undefined) { this.direction = null; }
            if (this.color === undefined) { this.color = null; }
            this.direction = direction;
            this.color = color;
        }

        public direction: RealVector;

        color: Color;

        public getColor(): string {
            return this.color.toWebString();
        }

        public getDirection(): number[] {
            return [this.direction.x, this.direction.y, this.direction.z];
        }
    }
    DirectionalLight["__class"] = "com.vzome.core.viewing.Lights.DirectionalLight";

}
